File: app/src/main/java/com/filemanager/fileexplorer/myfile/cleanmaster/free/adapter/GirlAdapter/GridAdapter.java
package com.filemanager.fileexplorer.myfile.cleanmaster.free.adapter.GirlAdapter;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.ThumbnailUtils;
import android.net.Uri;
import android.os.AsyncTask;
import android.provider.MediaStore;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import com.demo.example.R;
import com.filemanager.fileexplorer.myfile.cleanmaster.free.Ultil.Data_Manager;

import java.io.File;
import java.util.ArrayList;
import java.util.List;


public class GridAdapter extends RecyclerView.Adapter<GridViewHolder> {
    Context context;
    private Data_Manager dataManager;
    private SparseBooleanArray selectionState = new SparseBooleanArray();

    public GridAdapter(Data_Manager data_Manager, Context context) {
        this.dataManager = data_Manager;
        this.context = context;
    }

    @Override
    public GridViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        return new GridViewHolder(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.card_girl, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(GridViewHolder gridViewHolder, int i) {
        if (this.dataManager.getName(i).length() > 12) {
            gridViewHolder.textView.setText(this.dataManager.getName(i).substring(0, 11));
        } else {
            gridViewHolder.textView.setText(this.dataManager.getName(i));
        }
        gridViewHolder.imageView.setImageResource(this.dataManager.getIconId(i));
        gridViewHolder.itemView.setActivated(this.selectionState.get(i, false));
        new BackImageLoading(i).execute(gridViewHolder);
        if (this.selectionState.get(i, false)) {
            gridViewHolder.img_check.setVisibility(View.VISIBLE);
        } else {
            gridViewHolder.img_check.setVisibility(View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return this.dataManager.name.size();
    }

    public void toggleSelection(int i) {
        if (this.selectionState.get(i, false)) {
            this.selectionState.delete(i);
        } else {
            this.selectionState.put(i, true);
        }
        notifyItemChanged(i);
    }

    public void clearSelection() {
        this.selectionState.clear();
        notifyDataSetChanged();
    }

    public int getSelectedItemCount() {
        return this.selectionState.size();
    }

    public List<File> getSelectedItemsFile() {
        ArrayList arrayList = new ArrayList();
        for (int i = 0; i < this.selectionState.size(); i++) {
            arrayList.add(this.dataManager.getFiles(this.selectionState.keyAt(i)));
        }
        return arrayList;
    }


    public class BackImageLoading extends AsyncTask<GridViewHolder, GridViewHolder, Object> {
        Bitmap bitmap;
        private int i;

        BackImageLoading(int i) {
            this.i = i;
        }


        @Override
        public Object doInBackground(GridViewHolder... gridViewHolderArr) {
            String lowerCase = GridAdapter.this.dataManager.getFiles(this.i).getPath().toLowerCase();
            if (lowerCase.contains(".jpg") || lowerCase.contains(".png") || lowerCase.contains(".jpeg") || lowerCase.contains(".gif") || lowerCase.contains(".webp")) {
                if (GridAdapter.this.dataManager.getFiles(this.i).length() > 120000) {
                    long j = 0;
                    Uri uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
                    Cursor query = GridAdapter.this.context.getContentResolver().query(uri, new String[]{"_data", "_id"}, null, null, null);
                    query.moveToFirst();
                    while (true) {
                        if (query.getString(query.getColumnIndexOrThrow("_data")).equals(GridAdapter.this.dataManager.getFiles(this.i).getPath())) {
                            j = query.getLong(query.getColumnIndexOrThrow("_id"));
                            break;
                        } else if (!query.moveToNext()) {
                            break;
                        }
                    }
                    this.bitmap = MediaStore.Images.Thumbnails.getThumbnail(GridAdapter.this.context.getContentResolver(), j, 1, null);
                }
                if (this.bitmap == null) {
                    this.bitmap = ThumbnailUtils.extractThumbnail(BitmapFactory.decodeFile(GridAdapter.this.dataManager.getFiles(this.i).getPath()), 192, 192);
                }
                publishProgress(gridViewHolderArr);
            }
            return null;
        }


        @Override
        public void onProgressUpdate(GridViewHolder... gridViewHolderArr) {
            gridViewHolderArr[0].imageView.setImageBitmap(this.bitmap);
        }
    }
}

File: app/src/main/java/com/demo/filemanager/adapter/GirlAdapter/GridAdapter.kt
package com.demo.filemanager.adapter.GirlAdapter

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.ThumbnailUtils
import android.net.Uri
import android.provider.MediaStore
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.demo.filemanager.R
import com.demo.filemanager.Ultil.Data_Manager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class GridAdapter(private val dataManager: Data_Manager, private val context: Context) : RecyclerView.Adapter<GridViewHolder>() {
    private val selectionState = SparseBooleanArray()

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): GridViewHolder {
        return GridViewHolder(LayoutInflater.from(viewGroup.context).inflate(R.layout.card_girl, viewGroup, false))
    }

    override fun onBindViewHolder(gridViewHolder: GridViewHolder, i: Int) {
        gridViewHolder.textView.text = if (dataManager.getName(i)!!.length > 12) dataManager!!.getName(i)!!.substring(0, 11) else dataManager.getName(i)
        gridViewHolder.imageView.setImageResource(dataManager.getIconId(i))
        gridViewHolder.itemView.isActivated = selectionState.get(i, false)
        BackImageLoading(i, gridViewHolder).load()
        gridViewHolder.img_check.visibility = if (selectionState.get(i, false)) View.VISIBLE else View.GONE
    }

    override fun getItemCount(): Int {
        return dataManager.name!!.size
    }

    fun toggleSelection(i: Int) {
        if (selectionState.get(i, false)) {
            selectionState.delete(i)
        } else {
            selectionState.put(i, true)
        }
        notifyItemChanged(i)
    }

    fun clearSelection() {
        selectionState.clear()
        notifyDataSetChanged()
    }

    fun getSelectedItemCount(): Int {
        return selectionState.size()
    }

    fun getSelectedItemsFile(): List<Unit> {
        return List(selectionState.size()) { i ->
            dataManager.getFiles(selectionState.keyAt(i))
        }
    }

    inner class BackImageLoading(private val i: Int, private val gridViewHolder: GridViewHolder) {
        private var bitmap: Bitmap? = null

        fun load() {
            GlobalScope.launch(Dispatchers.IO) {
                val lowerCase = dataManager.getFiles(i)!!.path.lowercase()
                if (listOf(".jpg", ".png", ".jpeg", ".gif", ".webp").any { lowerCase.contains(it) }) {
                    if (dataManager.getFiles(i)!!.length() > 120000) {
                        var id: Long = 0
                        val uri: Uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                        val cursor = context.contentResolver.query(uri, arrayOf("_data", "_id"), null, null, null)
                        cursor?.moveToFirst()
                        while (cursor != null && !cursor.isAfterLast) {
                            if (cursor.getString(cursor.getColumnIndexOrThrow("_data")) == dataManager.getFiles(i)?.path) {
                                id = cursor.getLong(cursor.getColumnIndexOrThrow("_id"))
                                break
                            } else {
                                cursor.moveToNext()
                            }
                        }
                        cursor?.close()
                        bitmap = MediaStore.Images.Thumbnails.getThumbnail(context.contentResolver, id, MediaStore.Images.Thumbnails.MINI_KIND, null)
                    }
                    if (bitmap == null) {
                        bitmap = ThumbnailUtils.extractThumbnail(BitmapFactory.decodeFile(
                            dataManager.getFiles(i)!!.path), 192, 192)
                    }
                    withContext(Dispatchers.Main) {
                        gridViewHolder.imageView.setImageBitmap(bitmap)
                    }
                }
            }
        }
    }
}


